### PR TITLE
Remove `find_map_relevant_impl`

### LINKED ISSUE #108895: remove `find_map_relevant_impl`
https://github.com/rust-lang/rust/blob/9b60e6c68ff0aabad9a0edd71898466886dbf6bb/compiler/rustc_middle/src/ty/trait_def.rs#L153-L158

This function iterates over all impls which may apply to `self_ty` and returns the first `Some`. Note that just because an impl is given to `f` it may not actually apply to the whole `trait_ref` (or deeply matches `self_ty`). Because of this it ends up being very easy to misuse. Let's go through a few incorrect uses of this function:

https://github.com/rust-lang/rust/blob/9b60e6c68ff0aabad9a0edd71898466886dbf6bb/src/librustdoc/passes/collect_intra_doc_links.rs#L738-L763

only looks at the first maybe applicable impl even if there are multiple, idk how exactly that function is used, but it should be wrong :sweat_smile: 

https://github.com/rust-lang/rust/blob/9b60e6c68ff0aabad9a0edd71898466886dbf6bb/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs#L2195-L2197

Checks whether another trait with the same path has an impl for `self_ty`. Note that this emits the error message if 2 versions of a crate are used, but the substs implement *neither* of the traits.

https://github.com/rust-lang/rust/blob/9b60e6c68ff0aabad9a0edd71898466886dbf6bb/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs#L1818

whatever this is doing :sweat_smile: but this should be usable to point to an impl which doesn't actually apply.

I would like to change all the uses of `find_map_relevant_impl` to instead use `for_each_relevant_impl`.

### PR DESCRIPTION
**Pull request by compiler-errors**
Fixes #108895

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs [hunk 1]
```diff
@@ -1836,57 +1835,61 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
...
-                                self.resolve_vars_if_possible(predicate)
-                                    .print(FmtPrinter::new_with_limit(
-                                        self.tcx,
-                                        Namespace::TypeNS,
-                                        rustc_session::Limit(5),
-                                    ))
-                                    .unwrap()
-                                    .into_buffer()
-                            )),
+            let secondary_span = (|| {
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:1838]: this is a stable "try" "block" :⁾

**Comment by compiler-errors** [comment]
oh yeah, @vincenzopalazzo said they wanted to see this PR.

**Review by b-naber — APPROVED** [review]
Looks great.

**Pull request merged**